packages/engine/src/behave-graph/nodes/Profiles/Engine/helper/componentHelper.ts

```diff
@@ -148,3 +157,89 @@ export function getComponentGetters() {
   }
   return getters
 }
+
+type State = {
+  systemUUID: SystemUUID
+}
+
+const initialState = (): State => ({
+  systemUUID: '' as SystemUUID
+})
+
+export function getComponentListeners() {
+  const getters: NodeDefinition[] = []
+  const skipped: string[] = []
+  for (const [componentName, component] of ComponentMap) {
+    if (skipComponents.includes(componentName)) {
+      skipped.push(componentName)
+      continue
+    }
+    const outputsockets = generateComponentNodeSchema(component, true)
+    if (Object.keys(outputsockets).length === 0) {
+      skipped.push(componentName)
+      continue
+    }
+
+    const node = makeEventNodeDefinition({
+      typeName: `engine/component/on ${componentName} change`,
+      category: NodeCategory.Event,
+      label: `on ${componentName} change`,
+      in: {
+        entity: 'entity'
+      },
+      out: {
+        entity: 'entity',
+        ...outputsockets
+      },
+      initialState: initialState(),
+      init: ({ read, write, commit, graph }) => {
+        const entity = Number.parseInt(read('entity')) as Entity
+        const valueOutputs = Object.entries(node.out)
+          .splice(1)
+          .filter(([output, type]) => type !== 'flow')
+        const flowOutputs = Object.entries(node.out)
+          .splice(1)
+          .filter(([output, type]) => type === 'flow')
+
+        let prevComponentValue = {}
+        const systemUUID = defineSystem({
+          uuid: `behave-graph-on-${componentName}-change` + uniqueId(),
+          insert: { with: InputSystemGroup },
+          execute: () => {
+            const componentValue = getComponent(entity, component)
+            if (typeof componentValue !== 'object') {
+              if (prevComponentValue === componentValue) return
+              const value = EnginetoNodetype(componentValue)
+              write(valueOutputs[valueOutputs.length - 1][0] as any, value)
+              commit(flowOutputs[flowOutputs.length - 1][0] as any)
+              prevComponentValue = structuredClone(componentValue)
```

Comment:
Doing it like this is going to be extremely inefficient. Please use a reactor.

Reply:
i dont think u can embed a reactor in a behave graph node definition?
i know we can define systems as we need, can we do the same for reactors as well?

Reply:
all rewritten with reactor based logic, turns out you can define UseEffects in a loop as well

Reply:
I'm still not seeing the reactor logic here